Documentation on task mapping additions

uranusjr opened the pull request:
None of these actually exist yet. Instead of implementing things and discuss what the UI should be, let’s describe the UI first (and decide if they make sense), and try to make the UI happen later.

These are the additions planned for 2.5 on AIP-42 (task mapping), namely `expand_kwargs()`, `map()`, filter (using Airflow's existing mechanism), and `zip()`. The examples for `expand_kwargs()` and `zip()` are pretty bad now. Suggestions are much welcomed.

uranusjr committed:
Documentation on task mapping additions

None of these actually exist yet. These are the additions planned for
2.5 on AIP-42 (task mapping), namely expand_kwargs(), map(), filter
(using Airflow's existing mechanism), and zip().

The examples for expand_kwargs() and zip() are pretty bad now.
Suggestions are much welcomed.

Let's discuss the interface in this patch's thread; once we settle on
the syntax and semantic, I'll start implementing them.

dstandish commented on docs/apache-airflow/concepts/dynamic-task-mapping.rst:
```diff
@@ -208,27 +210,156 @@ In this example you have a regular data delivery to an S3 bucket and want to app
+
+    # copy_kwargs and copy_files are implemented the same.
+
+This makes ``copy_files`` only expand against ``.txt``, ``.json``, and ``.yml`` files, while ignoring the rest.
```
```suggestion
This makes ``copy_files`` only expand against ``.json`` and ``.yml`` files, while ignoring the rest.
```

potiuk reviewed (approved):
Fantastic update to the docs of task mapping and capabilities!

dstandish commented on docs/apache-airflow/concepts/dynamic-task-mapping.rst:
```diff
@@ -208,27 +210,156 @@ In this example you have a regular data delivery to an S3 bucket and want to app
+
+#. The callable argument of ``map()`` (``create_copy_kwargs`` in the example) **must not** be a task, but a plain Python function. The transformation is as a part of the “pre-processing” of the downstream task (i.e. ``copy_files``), not a standalone task in the DAG.
+#. The callable always take exactly one positional argument. This function is called for each item in the iterable used for task-mapping, similar to how Python’s built-in ``map()`` works.
+#. Since the callable is executed as a part of the downstream task, you can use any existing techniques to write the task function. To mark a component as skipped, for example, you should raise ``AirflowSkipException``. Note that returning ``None`` **does not** work here.
```
what happens when you return None?

uranusjr replied:
It would set the value to None, e.g. if you have `def foo(x)` and return None in the pre-processing transform, `foo` would simply receive `x=None`. (Note that since you can’t push None to XCom—an unfortunate limitation inherited from long ago—`x` cannot be None without pre-processing.)

uranusjr committed:
Detect common user error passing a task to map()

The argument to map() must be a plain function, not a task. This hook
checks for that.
